simplelru/expirable_lru.go

```diff
@@ -0,0 +1,276 @@
+package simplelru
+
+import (
+	"container/list"
+	"sync"
+	"time"
+)
+
+// ExpirableLRU implements a thread safe LRU with expirable entries.
+type ExpirableLRU struct {
+	size       int
+	purgeEvery time.Duration
+	ttl        time.Duration
+	done       chan struct{}
+	onEvicted  EvictCallback
+
+	sync.Mutex
+	items     map[interface{}]*list.Element
+	evictList *list.List
```

Comment:
Rather than using a list where you have to keep scanning through, I think you'd be better off using a priority queue. See https://github.com/hashicorp/vault/tree/master/sdk/queue and the Golang doc it links to. That way you simply add items to the queue where the sorting is by expiration time and only have to keep track of the first item in the queue in terms of expiration. This would also let you get rid of Close; a simple one-off timer would be the only goroutine you need to keep checking the first element.

Reply:
@jefferai, thanks for your thoughts! I understand the benefits of using a heap, but I don't understand how it would change anything.

removeElement is constant complexity now, it stays the same.

DeleteExpired complexity stays the same, as to delete all expired elements we either need to traverse whole dataset (`O(N)`) or go trough sorted list until we find first non-expired element (`O(N*logN)`) or check top element of the heap until we find a non-expiring element, which I believe would be exact same complexity as sorted list, as single heap rebalance would be `O(logN)` multiplied by N elements we want to delete (e.g. whole dataset during lifetime of the program).

If we want to have deleteExpired call in a goroutine, we need ability to call `Close()` to kill it. If I understand you correctly and you want me to not to call deleteExpired by timer, I don't understand which alternative you propose: even if we set one-time timer on new entry which will wait for it to expire and then clean it up, or clean up everything expired so far, we still want the ability to kill it explicitly.

Would you be so kind as to elaborate on what you want to see changed in that code behavior and what's the benefit of using heap for it?

Reply:
I think it's better to split this discussion into two separate threads: is there any way to get rid of the Close method (which is only possible if we won't have any goroutines spawned as I think) and is there a way to make this code faster.